numpy/_core/src/multiarray/arraywrap.c

```diff
@@ -0,0 +1,308 @@
+                }
+                continue;
+            }
+            double curr_priority = PyArray_GetPriority(obj, 0);
+            if (wrap == NULL || priority < curr_priority
+                    /* Prefer subclasses `__array_wrap__`: */
+                    || (curr_priority == 0 && wrap == Py_None)) {
+                Py_XSETREF(wrap, new_wrap);
+                Py_INCREF(Py_TYPE(obj));
+                Py_XSETREF(wrap_type, (PyObject *)Py_TYPE(obj));
+                priority = curr_priority;
+            }
+        }
+    }
+
+    if (wrap == NULL) {
+        Py_INCREF(Py_None);
+        wrap = Py_None;
+    }
+    if (wrap_type == NULL) {
+        Py_INCREF(&PyArray_Type);
+        wrap_type = (PyObject *)&PyArray_Type;
+    }
+
+    *out_wrap = wrap;
+    *out_wrap_type = wrap_type;
+
+    return 0;
+
+  fail:
+    Py_XDECREF(wrap);
+    Py_XDECREF(wrap_type);
+    return -1;
+}
+
+
+/* Get the arg tuple to pass in the context argument to __array_wrap__.
+ *
+ * Output arguments are only passed if at least one is non-None.
+ */
+static PyObject *
+_get_wrap_prepare_args(NpyUFuncContext *context) {
+    if (context->out == NULL) {
+        Py_INCREF(context->in);
+        return context->in;
+    }
+    else {
+        return PySequence_Concat(context->in, context->out);
+    }
+}
+
+
+/*
+ * Apply the array wrapping to a result array.
+ *
+ * @param obj The object (should be an array) to wrap.
+ * @param original_out NULL/None (both valid) or an object who's wrapping
+ *        method is always used/preferred.  The naming comes because for an
+ *        `out=` argument we always trigger its own wrapping method.
+ * @param wrap The array wrap function to call
+ * @param wrap_type The type belonging to the wrap function, when it matches
+ *        wrapping may be short-cut.
+ * @param context ufunc context or NULL, only used by normal ufunc calls.
+ * @param return_scalar Whether to prefer a scalar return.  Ignored when
+ *        `original_out` is passed, as an `out` argument is never scalar.
+ * @param force_wrap If True, we call the wrap (for subclasses) always,
+ *        because the ufunc would have mutated the content.
+ */
+NPY_NO_EXPORT PyObject *
+npy_apply_wrap(
+        PyObject *obj, PyObject *original_out,
+        PyObject *wrap, PyObject *wrap_type,
+        NpyUFuncContext *context, npy_bool return_scalar, npy_bool force_wrap)
+{
+    PyObject *res = NULL;
+    PyObject *new_wrap = NULL;
+    PyArrayObject *arr = NULL;
+    PyObject *err_type, *err_value, *traceback;
+
+    /* If provided, we prefer the actual out objects wrap: */
+    if (original_out != NULL && original_out != Py_None) {
+        /* 
+         * If an original output object was passed, wrapping shouldn't
+         * change it.  In particular, it doesn't make sense to convert to
+         * scalar.
+         */
+        return_scalar = NPY_FALSE;
+
+        if (PyArray_CheckExact(original_out)) {
+            wrap = Py_None;
```

Comment:
I stared at this for a long time, and thought that surely this needed `Py_INCREF(Py_None)`, but then I guess this is just a signal, and you leave `DECREF`ing wrap to the caller (logically), so this is fine. Maybe add a comment `/* Used as signal; wrap is not DECREF'd, so no need to INCREF(Py_None) here. */`